HPCC-33809 RoxiePipe trace context propagation

rpastrana opened the pull request:
- Adds new trace parent option (-tp)
- Injects traceparent to outgoing roxie bound request
- Adds new roxiepipetraced ecl testing case

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana committed:
HPCC-33809 RoxiePipe trace context propagation

- Adds new trace parent option (-tp)
- Injects traceparent to outgoing roxie bound request
- Adds new roxiepipetraced ecl testing case

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana commented:
Resulting span on Roxie side inherits parent from provided trace/span context:
![image](https://github.com/user-attachments/assets/f1aec368-9824-4af8-892a-aea7db0675d6)


The "run_workunit" span tracks hthor portion of the job, not sure why there's multiple roxie_echo, but they are treated as children of the root span as expected. The immediate question raised by the trace, are the wide gaps between the start of the run_workunit and the roxie_echo, the end of the graph1 span under each roxie_echo and the roxie_echo end, and the end of the last roxie_echo and the end of the run_workunit 

```
ecl run hthor /home/ubuntu/GIT/HPCC-Platform/testing/regress/ecl/roxiepipetraced.ecl
```


<Result>
<Dataset name='Result 1'>
 <Row><name><first>Joe       </first><last>Doe            </last></name><address><city>Fresno    </city><state>CA</state><zipcode>11111</zipcode></address></Row>
...
 <Row><name><first>Jane      </first><last>Smith          </last></name><address><city>Smallville</city><state>KY</state><zipcode>99999</zipcode></address></Row>
</Dataset>
</Result>
```

ghalliday reviewed (commented):
@rpastrana looks good.  one trivial comment, and one request for clarification.

rpastrana committed:
HPCC-33809 code review 1

- Removes nullptr test
- Clarifies comments on roxiepipetraced.ecl
- Outputs current trace information

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana commented on testing/regress/ecl/roxiepipetraced.ecl:
```diff
@@ -0,0 +1,84 @@
+    ' -h ' + TargetIP + ':9876 ' +
+    ' -r Peeps' +
+    ' -tp ' + lib_logging.Logging.getTraceSpanHeader() +  //Fetches current active trace/span context
+                                                          //Not currently supported on thor - run on hthor
```
Comment updated inline

ghalliday commented on testing/regress/ecl/roxiepipetraced.ecl:
```diff
@@ -0,0 +1,86 @@
+/*##############################################################################
```
This needs a corresponding xml file - see the failure from the regression suite.  
Also/alternatively, because the trace context will be different each time you may need to either remove that output or mark this test as not having a key file.

ghalliday reviewed (changes requested):
Minor- updates needed so regression test passes.
Also, please update the conclusion in the jira with what was implemented, and create a documentation jira to ensure it gets added to documentation/we circulate information about how to use it.

rpastrana committed:
HPCC-33809 code review 2

- Excludes test case from regression suite
- Removes nohthor nothor directives

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana commented:
> Resulting span on Roxie side inherits parent from provided trace/span context: ![image](https://private-user-images.githubusercontent.com/915949/432845649-f1aec368-9824-4af8-892a-aea7db0675d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4OTczNDUsIm5iZiI6MTc0NDg5NzA0NSwicGF0aCI6Ii85MTU5NDkvNDMyODQ1NjQ5LWYxYWVjMzY4LTk4MjQtNGFmOC04OTJhLWFlYTdkYjA2NzVkNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxN1QxMzM3MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zOTM4YjI3MTY5MDExM2U5OGU0ZTQ5YjA5N2MzYjJiMzIxNjY3ZjBhNDdlZmFiOGRmZTE1NzU5MjJmMzljNzYxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.On2RD7SK4cfmNdv7NRZD62pynbF6h7srR2kX_gxKsLc)
> 
> The "run_workunit" span tracks hthor portion of the job, not sure why there's multiple roxie_echo, but they are treated as children of the root span as expected. The immediate question raised by the trace, are the wide gaps between the start of the run_workunit and the roxie_echo, the end of the graph1 span under each roxie_echo and the roxie_echo end, and the end of the last roxie_echo and the end of the run_workunit
> 
> ```
> ecl run hthor /home/ubuntu/GIT/HPCC-Platform/testing/regress/ecl/roxiepipetraced.ecl
> ```
> 
> Joe Doe Fresno CA11111 ... Jane Smith SmallvilleKY99999 ```



> Resulting span on Roxie side inherits parent from provided trace/span context: ![image](https://private-user-images.githubusercontent.com/915949/432845649-f1aec368-9824-4af8-892a-aea7db0675d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4OTczNDUsIm5iZiI6MTc0NDg5NzA0NSwicGF0aCI6Ii85MTU5NDkvNDMyODQ1NjQ5LWYxYWVjMzY4LTk4MjQtNGFmOC04OTJhLWFlYTdkYjA2NzVkNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxN1QxMzM3MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zOTM4YjI3MTY5MDExM2U5OGU0ZTQ5YjA5N2MzYjJiMzIxNjY3ZjBhNDdlZmFiOGRmZTE1NzU5MjJmMzljNzYxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.On2RD7SK4cfmNdv7NRZD62pynbF6h7srR2kX_gxKsLc)
> 
> The "run_workunit" span tracks hthor portion of the job, not sure why there's multiple roxie_echo, but they are treated as children of the root span as expected. The immediate question raised by the trace, are the wide gaps between the start of the run_workunit and the roxie_echo, the end of the graph1 span under each roxie_echo and the roxie_echo end, and the end of the last roxie_echo and the end of the run_workunit

Following up on this conversation we had regarding this chart, the 3 roxie_echo sub-spans do seem to correlate to the number of "batches" expected, based on the -b option of "3" for the sample dataset of 9 records. My concern is the number of subspans for larger datasets - can we ensure appropriate batch sizes?
